Fatras/src/Digitization/PlanarSurfaceDrift.cpp

```diff
@@ -21,30 +23,30 @@ ActsFatras::PlanarSurfaceDrift::toReadout(const Acts::GeometryContext& gctx,
                                           const Acts::Vector3& driftDir) const {
   // Transform the hit & direction into the local surface frame
   const auto& invTransform = surface.transform(gctx).inverse();
-  Acts::Vector2 pos2Local = (invTransform * pos).segment<2>(0);
-  Acts::Vector3 seg3Local = invTransform.linear() * dir;
-  // Scale unit direction to the actual segment in the (depletion/drift) zone
-  seg3Local *= thickness / std::cos(Acts::VectorHelpers::theta(seg3Local));
-  // Calculate local entry/exit before drift
-  Acts::Vector2 entry = pos2Local - 0.5 * seg3Local.segment<2>(0);
-  Acts::Vector2 exit = pos2Local + 0.5 * seg3Local.segment<2>(0);
-  // Actually apply a drift
-  // - dirftDir is assumed in local coordinates
-  if (!driftDir.segment<2>(0).isApprox(Acts::Vector2(0., 0.))) {
-    // Apply the scaled drift
-    auto applyDrift = [&](Acts::Vector2& local) {
-      auto scaledDriftDir =
-          driftDir * thickness / std::cos(Acts::VectorHelpers::theta(driftDir));
-      local += scaledDriftDir.segment<2>(0);
-    };
-
-    if (driftDir.z() > 0.) {
-      applyDrift(entry);
-    }
-    if (driftDir.z() < 0.) {
-      applyDrift(exit);
-    }
+  Acts::Vector3 pos3Local(invTransform * pos);
+  Acts::Vector3 seg3Local = invTransform.linear() * dir.normalized();
+  if (std::abs(seg3Local.z()) < Acts::s_epsilon) {
+    // The segment is parallel to the surface
+    return ActsFatras::DigitizationError::DriftError;
   }
-
-  return {entry, exit};
+  // Scale the unit vector to the thickness of the module
+  double scale = thickness / seg3Local.z();
+  seg3Local *= scale;
+  // The drift direction is in the local frame, so we need to transform it
+  Acts::Vector3 entry = pos3Local - 0.5 * seg3Local;
+  Acts::Vector3 exit = pos3Local + 0.5 * seg3Local;
+  Acts::Vector3 driftedEntry = entry;
+  Acts::Vector3 driftedExit = exit;
+  // Apply the drift if it is not zero
+  if (Acts::VectorHelpers::perp(driftDir) > Acts::s_epsilon &&
+      std::abs(driftDir.z()) > Acts::s_epsilon) {
+    // Apply the drift to the entry and exit points
+    double driftScale = 0.5 * thickness / driftDir.z();
+    driftedEntry += driftScale * driftDir;
+    driftedExit -= driftScale * driftDir;
+  }
+  // The drifted segment and the original segment
+  return std::tuple<Segment2D, Segment3D>{
+      Segment2D{driftedEntry.segment<2>(0), driftedExit.segment<2>(0)},
+      Segment3D{entry, exit}};
```

Comment:
```suggestion
  return std::tuple{
      Segment2D{driftedEntry.segment<2>(0), driftedExit.segment<2>(0)},
      Segment3D{entry, exit}};
```

Should be able to use CTAD here.

Reply:
Changed.